File: feature_job_generator.py
import os
import re
from collections import namedtuple

def remove_option(x):
  return x.replace("Option[", "").replace("]", "")

def lower_plural(t):
  return t.lower() + "s"

class FeatureSet:
  def __init__(self, name, set_type, type_params):
    self.name = name
    self.set_type = set_type
    self.type_params = type_params
    self.data_source_template = "  val %(val)-14s = HiveTextSource[%(type)s, Nothing](new Path(\"data/%(dir)s\"), partitions%(delim)s)"
    self.delims = { "Movie": "", "User": "", "Rating": ", \"\\t\"" }
  def split_types(self):
    return self.type_params.translate(None, "()").split(", ")
  def thrift_imports(self):
    types = map(remove_option, self.split_types())
    if len(types) == 1:
      return types[0]
    else:
      return "{{{}}}".format(", ".join(types))
  def data_sources(self):
    types = self.split_types()
    def map_data_sources(t):
      tx = remove_option(t)
      return self.data_source_template % { "val": lower_plural(tx), "type": tx, "dir": lower_plural(tx), "delim": self.delims[tx] }
    return "\n".join(map(map_data_sources, types))
  def binding(self):
    def val_names(types):
      return map(lower_plural, map(remove_option, types))
    types = self.split_types()
    type_str = ", ".join(val_names(types))
    if len(types) == 1:
      return "from(" + type_str + ")"
    elif "Option[" in types[1]:
      return "leftJoin(" + type_str + ")"
    else:
      return "join(" + type_str + ")"
  def table_name(self):
    return lower_plural(remove_option(self.split_types()[0]))


feature_set_pattern = "(\w+) extends (\w*FeatureSet\w*)\[((?:\([\w\,\s]+\))|(?:\w+))"
feature_job_pattern = "(\w+)Job extends \w+Job"

jobs = []

source_dir = './examples/target/scala-2.11/src_managed/main'
dest_dir = './examples/src/main/scala/commbank/coppersmith/examples'

files = [os.path.join(source_dir,f) for f in os.listdir(source_dir) if os.path.isfile(os.path.join(source_dir,f))]

for f in files:
  for line in open(f):
    feature_job = re.search(feature_job_pattern, line)
    if (feature_job is not None):
      jobs.append(feature_job.group(1))

sets = []
for f in files:
  for line in open(f):
    feature_set = re.search(feature_set_pattern, line)
    if (feature_set is not None and feature_set.group(1) not in jobs):
      sets.append(FeatureSet(feature_set.group(1), feature_set.group(2), feature_set.group(3)))


template = """package commbank.coppersmith.examples.userguide

import org.apache.hadoop.fs.Path

import com.twitter.scalding.Config

import org.joda.time.DateTime

import au.com.cba.omnia.maestro.api.Maestro.DerivedDecode

import commbank.coppersmith.api._, scalding._
import commbank.coppersmith.examples.thrift.%(thrift_imports)s

case class %(name)sConfig(conf: Config) extends FeatureJobConfig[%(type_params)s] {
  val partitions     = ScaldingDataSource.Partitions.unpartitioned
%(data_sources)s

  val featureSource  = %(name)s.source.bind(%(binding)s)

  val featureContext = ExplicitGenerationTime(new DateTime(2015, 1, 1, 0, 0))

  val featureSink    = EavtSink.configure("userguide", new Path("dev"), "%(table_name)s")
}

object %(name)sJob extends SimpleFeatureJob {
  def job = generate(%(name)sConfig(_), %(name)s)
}
"""

for s in sets:
  out = template % { 'thrift_imports': s.thrift_imports(),
               'name': s.name,
               'type_params': s.type_params,
               'data_sources': s.data_sources(),
               'binding': s.binding(),
               'table_name': s.table_name()
             }
  with open(os.path.join(dest_dir, s.name + "Job.scala"), "w+") as f:
    f.write(out)

File: project/FeatureJobGenerator.scala
import sbt._
import scala.io.Source

object FeatureJobGenerator {
  def gen(files: Seq[File]): Seq[(String, String)] = {
    constructors(files) map { f => f.name -> f.construct }
  }

  class FeatureJobConstructor(val name: String, typeParams: String) {
    val delims = Map("Rating" -> """\t""")
    val splitTypes = typeParams.replaceAll("[()]", "").split(", ").toList

    def dataSourceString(collectionName: String, typeParam: String, delim: Option[String]) = {
      val delimString = (delim map (d => s""", \"$d\"""")).getOrElse("")
      f"val $collectionName%-14s = HiveTextSource[$typeParam, Nothing]" +
        f"""(new Path(\"data/$collectionName\"), partitions$delimString)"""
    }

    def removeOption(t: String) =
      t.replace("Option[", "").replace("]", "")

    def toLowerPlural(s: String) =
      s.toLowerCase + "s"

    val collectionNames: List[String] = splitTypes map (removeOption _ andThen toLowerPlural)

    def tableName: String = collectionNames match {
      case n :: _ => n
      case _ => ""
    }

    val thriftImport = {
      val importTypes = splitTypes map removeOption
      importTypes match {
        case List(single) => single
        case imports       => imports.mkString("{", ", ", "}")
      }
    }

    val dataSources = {
      (for {
        t <- splitTypes map removeOption
      } yield dataSourceString(toLowerPlural(t), t, delims.get(t))).mkString("\n  ")
    }

    val binding = {
      val dataSources = collectionNames.mkString(", ")
      if (splitTypes.length == 1) s"from($dataSources)"
      else if (splitTypes(1).contains("Option[")) s"leftJoin($dataSources)"
      else s"join($dataSources)"
    }

    val construct: String = {
      s"""|package commbank.coppersmith.examples.userguide
          |
          |import org.apache.hadoop.fs.Path
          |
          |import com.twitter.scalding.Config
          |
          |import org.joda.time.DateTime
          |
          |import au.com.cba.omnia.maestro.api.Maestro.DerivedDecode
          |
          |import commbank.coppersmith.api._, scalding._
          |import commbank.coppersmith.examples.thrift.$thriftImport
          |
          |case class ${name}Config(conf: Config) extends FeatureJobConfig[$typeParams] {
          |  val partitions     = ScaldingDataSource.Partitions.unpartitioned
          |  $dataSources
          |
          |  val featureSource  = $name.source.bind($binding)
          |
          |  val featureContext = ExplicitGenerationTime(new DateTime(2015, 1, 1, 0, 0))
          |
          |  val featureSink    = EavtSink.configure("userguide", new Path("dev"), "$tableName")
          |}
          |
          |object ${name}Job extends SimpleFeatureJob {
          |  def job = generate(${name}Config(_), $name)
          |}
          |""".stripMargin
    }
  }

  val featureSetPattern = """(\w+) extends \w*FeatureSet\w*\[((?:\([\w\,\s\[\]]+\))|(?:\w+))""".r.unanchored
  val jobPattern = """(\w+)Job extends \w+Job""".r.unanchored

  def existingJobs(files: Seq[File]): Seq[String] =
    for {
      file <- files
      line <- Source.fromFile(file).getLines()
      List(job) <- jobPattern.unapplySeq(line)
    } yield job

  def constructors(files: Seq[File]): Seq[FeatureJobConstructor] = {
    val existing = existingJobs(files)
    for {
      file <- files
      line <- Source.fromFile(file).getLines()
      List(name, typeParams) <- featureSetPattern.unapplySeq(line)
      if !existing.contains(name)
    } yield new FeatureJobConstructor(name, typeParams)
  }


}